task-launcher/src/tasks/hearts-and-flowers/timeline.js

```diff
@@ -329,15 +323,30 @@ function getMixedPracticeSection(adminConfig) {
   // flower-practice-feedback2, "When you see a FLOWER, press the button on the OPPOSITE side."
   const practiceFeedback = buildMixedPracticeFeedback('heartPracticeFeedback2', 'feedbackGoodJob', 'flowerPracticeFeedback2', 'feedbackGoodJob');
 
+  // Let's prepare a callback to pass to our stimuli trials and keep track of successive correct practice trials
+  let practiceWinStreakCount = 0;
+  function onTrialFinishTimelineCallback(data) {
+    practiceWinStreakCount = data.correct ? practiceWinStreakCount+1 : 0;
+    console.log(`Practice block win streak=${practiceWinStreakCount}`);
+    if (practiceWinStreakCount >= adminConfig.correctPracticeTrial) {
+      console.log(`Ending practice block early: win streak=${practiceWinStreakCount}`);
+      jsPsych.endCurrentTimeline();
+    }
+  };
+
```

Comment:
Why do we need this block defined twice? Also can you remove the logs please.

Reply:
The adminConfig object is a different one and the logic is used only twice. I thought about introducing a function "getOnTrialFinishTimelineCallback(adminConfig)" that both mixed and non-mixed practice logic would call, but doing so would hide the implementation details (particularly the fact that the callback takes a parameter). Ideally we may want to create a wrapper class/object to abstract away this quirky behavior which need to be scoped to the timeline level rather than trial level.

Reply:
I may be wrong on the approach though. It's just that so far I could not see a good way to implement behaviors that span a particular nested timeline. You can pass a on_finish mapping here (as commented above circa line 260 - 250) but it gets completely overridden if you also pass one with your "trial" object.

Reply:
I think the "jsPsych way" is either to leverage the on_trial_finish callback at the experiment level or to use [conditional trials (by passing a conditional function)](https://www.jspsych.org/6.3/overview/timeline/#conditional-timelines). But both method still get you in a bit of an awkward situation (where you are either scoped to the experiment or the trial but need to apply logic on a specific nested timeline).